File: sdk/ios/Sources/DuallerSDK/RequestHandler.swift
import Foundation

/// Handles HTTP/network requests from the mini-program
public final class RequestHandler: Handler {
    public let apiName: String = "dualler.request"

    public func handle(_ message: BridgeMessage) -> String? {
        guard let payload = message.payload,
              let urlString = payload["url"] as? String else {
            return "{\"errMsg\":\"invalid request URL\"}"
        }

        guard let url = URL(string: urlString) else {
            return "{\"errMsg\":\"invalid URL\"}"
        }

        var request = URLRequest(url: url)
        if let method = payload["method"] as? String {
            request.httpMethod = method
        }
        if let data = payload["data"] as? String {
            request.httpBody = data.data(using: .utf8)
        }

        let semaphore = DispatchSemaphore(value: 0)
        var resultJSON: String?
        var errorDesc: String?

        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                errorDesc = error.localizedDescription
                semaphore.signal()
                return
            }

            if let httpResponse = response as? HTTPURLResponse {
                let statusCode = httpResponse.statusCode
                let body = (data?.map { String(describing: $0) }) ?? ""
                resultJSON = String(format: "{\"statusCode\":%d,\"data\":\"%s\"}", statusCode, body)
            }
            semaphore.signal()
        }.resume()

        semaphore.wait()

        if let desc = errorDesc {
            return "{\"errMsg\":\"request failed: \(desc)\"}"
        }
        return resultJSON ?? "{\"errMsg\":\"request unknown error\"}"
    }
}

File: sdk/android/src/main/java/com/dualler/sdk/handler/RequestHandler.kt
package com.dualler.sdk.handler

import com.dualler.sdk.protocol.APIContract
import com.dualler.sdk.protocol.BridgeMessage
import java.net.HttpURLConnection
import java.net.URL

class RequestHandler : Handler {
    override val apiName: String = APIContract.REQUEST

    override fun handle(message: BridgeMessage): String? {
        val payload = message.payload as? Map<*, *> ?: return null
        val urlStr = payload["url"] as? String ?: return null
        val method = (payload["method"] as? String) ?: "GET"

        return try {
            val conn = URL(urlStr).openConnection() as HttpURLConnection
            conn.requestMethod = method
            conn.connect()

            val body = conn.inputStream.readBytes().decodeToString()
            conn.disconnect()

            """{"statusCode":${conn.responseCode},"data":$body}"""
        } catch (e: Exception) {
            """{"errMsg":"request fail: ${e.message}"}"""
        }
    }
}
